### PR TITLE
[ESIMD] Add/fix tests for (slm_)gather/scatter.

### PR DESCRIPTION
**Pull request by kbobrovs**
This is a complementary PR for https://github.com/intel/llvm/pull/4771.

Signed-off-by: Konstantin S Bobrovsky <konstantin.s.bobrovsky@intel.com>

### CODE REVIEW

File: SYCL/ESIMD/api/slm_gather_scatter.cpp [hunk 1]
```diff
@@ -118,12 +128,11 @@ int main(void) {
 
   bool passed = true;
 
-#ifndef SKIP_ONE_AND_TWO_BYTE_BLOCKS
   passed &= test<char, 16>(q);
   passed &= test<char, 32>(q);
   passed &= test<short, 16>(q);
   passed &= test<short, 32>(q);
-#endif
+
```
Comment (v-klochkov) [inline SYCL/ESIMD/api/slm_gather_scatter.cpp:135]: The patch in headers enabled the other sizes as well: 1 and 8:
https://github.com/intel/llvm/pull/4771/files#diff-5a799b2d2deec957626a491f9bb2fdd28cb52ac9a9ecba3c6ddd87e732c8aa7fR729
Wouldn't it be useful to test them as well? )Not necessary in this patch, but in separate one)

File: SYCL/ESIMD/regression/variable_gather_mask.cpp [hunk 2]
```diff
@@ -0,0 +1,161 @@
...
+// Part of the LLVM Project, under the Apache License v2.0 with LLVM Exceptions.
+// See https://llvm.org/LICENSE.txt for license information.
+// SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+//
+//===----------------------------------------------------------------------===//
+// REQUIRES: gpu
+// UNSUPPORTED: cuda || hip
+// RUN: %clangxx -fsycl %s -o %t.out
+// RUN: %GPU_RUN_PLACEHOLDER %t.out
+// XFAIL: *
```
Comment (kbobrovs) [inline SYCL/ESIMD/regression/variable_gather_mask.cpp:12]:
```suggestion
// XFAIL: *
// TODO: once BE problem is fixed and this test starts to pass, need to also
// enable TEST_VECTOR_VAR_MASK test cases in slm_gather_scatter_heavy.cpp
```

File: SYCL/ESIMD/api/slm_gather_scatter_heavy.cpp [hunk 3]
```diff
@@ -0,0 +1,393 @@
...
+      constexpr auto WG_SIZE = STRIDE; // for simplicity of the test
+      KernelType kernel(acc_in, acc_out, MaskedLane);
+      cgh.parallel_for(nd_range<1>{glob_range, range<1>(WG_SIZE)}, kernel);
+    });
+    e.wait();
+  } catch (cl::sycl::exception const &e) {
+    std::cout << "SYCL exception caught: " << e.what() << '\n';
+    delete[] A;
+    delete[] B;
+    return e.get_cl_code();
```
Comment (v-klochkov) [inline SYCL/ESIMD/api/slm_gather_scatter_heavy.cpp]: get_cl_code() is obsolete API causing warnings.
```suggestion
    return false;
```

File: SYCL/ESIMD/api/slm_gather_scatter.cpp (hunk 1 above)
Comment (kbobrovs) [reply SYCL/ESIMD/api/slm_gather_scatter.cpp:135]: Good point. slm_scalar_load/slm_scalar_store actually do that (slm_scalar*<T>(...) = slm*<T, 1>(...)). But I agree there should be a test case directly testing length 1.

File: SYCL/ESIMD/api/slm_gather_scatter_heavy.cpp (hunk 3 above)
Comment (kbobrovs) [reply SYCL/ESIMD/api/slm_gather_scatter_heavy.cpp]: good catch!

**Review by v-klochkov — DISMISSED** [review]

**Review by v-klochkov — APPROVED** [review]

**Comment by kbobrovs** [comment]
@vladimirlaz, could you please merge?
The two failures
[2021-10-19T01:50:34.806Z]   SYCL :: ESIMD/api/slm_gather_scatter.cpp
[2021-10-19T01:50:34.806Z]   SYCL :: ESIMD/api/slm_gather_scatter_heavy.cpp
are expected and will go away once new compiler with intel/llvm#4771 is used

**Pull request merged**